src/lib/array.js

```diff
@@ -63,6 +65,112 @@ exports.ensureArray = function(out, n) {
+    uint8: typeof Uint8Array !== 'undefined' ? Uint8Array : null,
+    uint8clamped: typeof Uint8ClampedArray !== 'undefined' ? Uint8ClampedArray : null,
+    int16: typeof Int16Array !== 'undefined' ? Int16Array : null,
+    uint16: typeof Uint16Array !== 'undefined' ? Uint16Array : null,
+    int32: typeof Int32Array !== 'undefined' ? Int32Array : null,
+    uint32: typeof Uint32Array !== 'undefined' ? Uint32Array : null,
+    float32: typeof Float32Array !== 'undefined' ? Float32Array : null,
+    float64: typeof Float64Array !== 'undefined' ? Float64Array : null,
+};
+exports.typedArrays = typedArrays;
+
+
+exports.decodeTypedArraySpec = function(v) {
+    // Assume processed by coerceTypedArraySpec
+    v = coerceTypedArraySpec(v);
+    var T = typedArrays[v.dtype];
+    var buffer;
+    if(v.bvals.constructor === ArrayBuffer) {
+        // Already an ArrayBuffer
+        buffer = v.bvals;
+    } else {
+        // Decode, assuming a string
+        buffer = b64.decode(v.bvals);
+    }
+
+    // Check if 1d shape. If so, we're done
+    if(v.ndims === 1) {
+        // Construct single Typed array over entire buffer
+        return new T(buffer);
+    } else {
+        // Reshape into nested plain arrays with innermost
+        // level containing typed arrays
+        // We could eventually adopt an ndarray library
+
+        // Build cumulative product of dimensions
+        var cumulativeShape = v.shape.map(function(a, i) {
+            return a * (v.shape[i - 1] || 1);
+        });
+
+        // Loop of dimensions in reverse order
+        var nestedArray = [];
+        for(var dimInd = v.ndims - 1; dimInd > 0; dimInd--) {
+            var subArrayLength = v.shape[dimInd];
+            var numSubArrays = cumulativeShape[dimInd - 1];
+            var nextArray = [];
+
+            if(dimInd === v.ndims - 1) {
+                // First time through, we build the
+                // inner most typed arrays
+                for(var typedInd = 0; typedInd < numSubArrays; typedInd++) {
+                    var typedOffset = typedInd * subArrayLength;
+                    nextArray.push(
+                        new T(buffer, typedOffset * T.BYTES_PER_ELEMENT, subArrayLength)
+                    );
+                }
+            } else {
+                // Following times through, build
+                // next layer of nested arrays
+                for(var i = 0; i < numSubArrays; i++) {
+                    var offset = i * subArrayLength;
+                    nextArray.push(nextArray.slice(offset, offset + subArrayLength - 1));
+                }
+            }
+
+            // Update nested array with next nesting level
+            nestedArray = nextArray;
+        }
+
+        return nestedArray;
+    }
+};
+
+function isTypedArraySpec(v) {
+    // Assume v has not passed through
+    return isPlainObject(v) && typedArrays[v.dtype] && v.bvals && (
+        Number.isInteger(v.shape) ||
```

Comment:
Looks like `Number.isInteger` isn't supported by IE... the test we use for coercing integers amounts to `isNumeric(val) && (val % 1 === 0)`. That also accepts integer strings, which AFAICT would work in the rest of the code, but perhaps simpler would be `Math.round(val) === val` which does not accept strings - also seems fine, I see no reason we need to allow strings here.

Reply:
Do we actually need `shape` for 1D arrays? I don't see it being used in that case, the b64 decoder seems to do that automatically.

Reply:
I don't get the use of allowing shape to be an integer, t.b.h. My inclination would be to simply require shape to be a list and avoid all the triaging. And then maybe support omitting the shape.

Reply:
I'm fine dropping shape as integer support. I think I'd like to keep it as a 1D list for 1D arrays for consistency for now

Reply:
Sure, we shouldn't prevent people from including the explicit 1D shape, even though we'll ultimately ignore it. But I think we should make it optional (and document it as such)